RFC: Implementing new mechanism for hard register operands to inline asm as a constraint.

tltao opened the pull request:
Recreate the hard register PR created on phabricator with some minor additions: https://reviews.llvm.org/D105142

The main idea is to allow Clang to support the ability to specify specific hardware registers in inline assembly constraints via the use of curly braces ``{}``. As such, this is mainly a Clang change.

Relevant RFCs posted here:

https://lists.llvm.org/pipermail/llvm-dev/2021-June/151370.html
https://gcc.gnu.org/pipermail/gcc/2021-June/236269.html (implemented - see https://gcc.gnu.org/onlinedocs/gcc/Hard-Register-Constraints.html)

Copying the Summary from the phabricator patch:


- This is put up as an RFC patch to get feedback about the introduction of a new inline asm constraint which supports hard register operands
- This is mostly a clang change, since the LLVM IR for the inline assembly already supports the {...} syntax which the backend recognizes. This change merely completes the loop in terms of introducing a user facing constraint which maps to it.

The following design decisions were taken for this patch:

For the Sema side:

- We validate the "{" constraint using a two-phase validation approach. Firstly, we check if there is a target dependent implementation to handle the {....} constraint. If it fails, then we move on to a target agnostic check where we parse and validate the {....} constraint.
- Why do we do this? Well, there are some targets which already process the {...} as a user facing constraint. For example the AMDGPU target. It supports syntax of the form {register-name} as well as {register-name[...]}. Moving this implementation to the target agnostic side seems to set a precedent, that we can keep extending the target agnostic implementation based on new cases for certain targets, which would be better served of moving it to the respective target.
- In terms of the target agnostic validation, we simply check for the following syntax {.*}. The parsed out content within the two curly braces is checked to see whether its a "valid GCC register".

For the Clang CodeGen side:

- Most of the work is done in the AddVariableConstraints function in CGStmt.cpp, which is responsible for emitting the LLVM IR corresponding to the usage of an actual register that the backend can use. Coincidentally, the LLVM IR is also {...}. As mentioned above, this is essentially mapping the LLVM Inline Assembly IR back to a user facing inline asm constraint.
- Within this function we add in logic to check if the constraint is of the form [&]{...} in addition to the "register asm" construct.
- A scenario where it was applicable to apply both the "register asm" construct and the "hard register inline asm constraint" was diagnosed as an unsupported error because there's no way the compiler will know which register the user meant. The safest option here is to error out explicitly, and put onus back on the user.
- To achieve this, and refactor it in a nice way, most of the logic pertaining to the "register asm" construct has been moved into a separate function called ShouldApplyRegisterVariableConstraint which deduces whether to apply the "register asm" construct or not.
- Furthermore, the GCCReg field is set with the "Register" only if the register is validated to be a "valid GCC Register" type. To me it doesn't make a lot of sense to set GCC Reg to something that might not necessarily be a "valid GCC register" as defined by respective targets. Would we have a case where we could have a {.*} constraint where the contents inside the curly brace is not a valid register? Yes. For example, The x86 target supports the "@cca" constraint, which is validated on the Sema side as a valid constraint, and before processing is converted to "{@cca}" (via the convertAsmConstraint function. "@cca" is not a valid "GCC register name". So in these case, we'll simply emit the constraint without setting GCCReg (with the assumption that the respective backends deal with it appropriately)

Tests:

- Various tests were updated to account for the new behaviour.
- I added a few SystemZ tests because we work on the Z backend, but I have no issues adding testing for multiple targets.
- There were a few mentions of "waiting" for the GCC implementation of the same to land before the Clang side lands. As mentioned above, the intent to implement it on the GCC side has already been put forward via the RFC. I have no issues "parking" this implementation until its ready to be merged in. However, it might be good to hash out any open questions/concerns in the interim.

Differences with GCC:

For the most part, the Clang implementation follows the GCC implementation. However, one difference between the Clang implementation and the GCC implementation is in the following example:

```
int x;
asm ("" : "=r" (x), "={r1}" (x));  (1)
asm ("" : "=r2" (x), "={r1}" (x));  (2)
```

In the GCC implementation, both statements above emits the error `matching constraint not valid in output operand`. In LLVM, statement `(1)` emits the error `couldn't allocate output register for constraint 'x'` from `InlineAsmLowering/selectionDAG` whereas statement `(2)` emits no errors.

jyknight commented:
Thanks for reviving the change! 

I'm definitely in favor of this, but since it adds no new functionality (only better usability), the value proposition is significantly lessened (IMO, to the point where it may not be be worthwhile to do) if it gets implemented ONLY in clang. So I'd love to see some movement on the GCC side, too. Maybe revive the existing discussion thread?

tltao commented on clang/lib/Basic/TargetInfo.cpp:
```diff
@@ -770,6 +770,18 @@ bool TargetInfo::validateOutputConstraint(ConstraintInfo &Info) const {
     case 'E':
     case 'F':
       break;  // Pass them.
+    case '{': {
```
Yes, we should be using the same names. The register asm variable is parsed in `clang/lib/Sema/SemaDecl.cpp` using:
```
case SC_Register:
  // Local Named register
  if (!Context.getTargetInfo().isValidGCCRegisterName(Label) &&
      DeclAttrsMatchCUDAMode(getLangOpts(), getCurFunctionDecl()))
    Diag(E->getExprLoc(), diag::err_asm_unknown_register_name) << Label;
  break;
```

tltao commented:
Ping. Restarting this topic as GCC have implemented the same feature. https://gcc.gnu.org/onlinedocs/gcc/Hard-Register-Constraints.html

efriedma-quic reviewed (commented):
Please add tests for multiple constraints, like `void f() { asm(""::"i{x0}"(0)); }`.

uweigand commented:
I can image cases where multiple register constraints may still be useful; e.g. to express that we can use any one of several hard-coded registers, or we can use a hard-coded register or some register class that doesn't contain that hard register.

It is of course superfluous to specific both a hard register and a register class containing this register.  But that's really the same issue as specifying two classes where one is a superclass of the other - and we don't explicitly forbid this either.

tltao commented:
Thanks for all the comments. I think from @stefan-sf-ibm's response, it sounds like the GCC side is not fully decided on what to do with multiple constraints involving hard registers, so we can wait for the details to be hashed out further there. 

In the meantime, I plan on making a few changes:

- Modify the current `register asm` implementation in Clang to append any constraints to the `register asm` register. This follows the GCC implementation. One caveat is there's no type checking in Clang for constraints, so we have no way of excluding the register asm register if there's a mismatch. However, because `TargetLowering::C_RegisterClass` has higher priority than `TargetLowering::C_Register`, we will pick the register class constraint instead of the register asm constraint in `TargetLowering`.
- Allow Clang to accept multiple constraints involving hard registers. As mentioned before, there is no real logic in `TargetLowering` to pick the "best" choice. It simply selects based on a fixed priority given to the different constraint types, and if all the constraints are the same priority, then it just picks the first one. I think it's outside the scope of this PR to do any work here, so I want to leave it as is.
- Add tests for these scenarios.

Does this sound reasonable?

tltao commented:
> I'm not sure I follow what you mean here; does this mean that if you have a register asm variable, it might not end up in that register? That seems like a regression for existing code.

Ah, good point. I was thinking more in line of mismatching register classes, but you're right, with our current TargetLowering logic, it's not possible to combine register asm with any other constraints.

tltao commented:
Hmm looks like the testing reveals another problem. We have no way of invalidating a platform specific constraint that also uses `{}`.

The AMDGCN tests expect `{exec}a` to fail to verify, but it does not fail in the generic `validateHardRegisterAsmConstraint` since we want to support something like `{r1}r`.

@AlexVlx I couldn't find any documentation on the usage of `{}` on AMDGPU, is failing the constraint `{exec}a` necessary? If so, we might have to also reject `{r1}r`.

Commit:
Revisiting the hard register constraint PR on phabricator:
https://reviews.llvm.org/D105142

The main idea is to allow Clang to support the ability to
specify specific hardware registers in inline assembly
constraints via the use of curly braces ``{}``. As such,
this is mainly a Clang change.

Relevant RFCs posted here

https://lists.llvm.org/pipermail/llvm-dev/2021-June/151370.html
https://gcc.gnu.org/pipermail/gcc/2021-June/236269.html

Commit:
Rebasing updates and disable multiple constraints of different types

Commit:
remove debug statement and add additional test case

Commit:
Formatting
